### PR TITLE
[Qt][WIP] allow possibility to add a comment to a WalletTx

### PR DESCRIPTION
**Pull request by jonasschnelli**
Try to sync rpc/qt wallets implementations.
Solves #2086.

bildschirmfoto 2015-03-15 um 16 40 49

bildschirmfoto 2015-03-15 um 16 40 59

```
    {
        "account" : "",
        "address" : "mm2STH7X3rbBVd73XaDrwdJHM69ukj31N8",
        "category" : "send",
        "amount" : -1.00000000,
        "vout" : 1,
        "fee" : -0.00000521,
        "confirmations" : 0,
        "txid" : "151e1e32090b765ebc30ebcc385bd79978081130b331972b94e142f3aeb85e10",
        "walletconflicts" : [
        ],
        "time" : 1426434053,
        "timereceived" : 1426434053,
        "comment" : "This is a test wtx comment."
    }
```

### CODE REVIEW

File: src/qt/walletmodeltransaction.cpp [hunk 1]
```diff
@@ -65,3 +65,8 @@ CReserveKey *WalletModelTransaction::getPossibleKeyChange()
 {
     return keyChange;
 }
+
+void WalletModelTransaction::setComment(QString comment)
```
Comment (Diapolo) [inline src/qt/walletmodeltransaction.cpp]: Can this be a const reference?

**Comment by laanwj** [comment]
Agree on the idea. Maybe we can make the comment editable too in the details view? As the information is purely local, there is no reason why it shouldn't be possible to add comments in retrospect.

**Comment by jonasschnelli** [comment]
@laanwj Yes. This is a good idea. Let me try to extend the editable comment to the transaction details window. What about adding a possibility to change the comment over RPC (without adding a new rpc call)?

**Comment by luke-jr** [comment]
How is this any different from the Label? Seems redundant.

**Comment by jonasschnelli** [comment]
@luke-jr comments already exists in RPC world and are per walletTx.
Labels are per address as far as I know.

**Comment by jonasschnelli** [comment]
@laanwj making the local comment editable within the TransactionDetailDialog looks after a big diff. It needs changes here and there and may be hard to review. It would also require changes within `CWallet`. IMO wtx are currently meant to be immutable. I would like to keep the PR as it is and address this "extension" in another PR.

**Comment by luke-jr** [comment]
@jonasschnelli Addresses are per walletTx generally as well (at least in proper use, which should be encouraged). RPC "comments" are basically just the deprecated accounting stuff which didn't have addresses associated.

**Comment by jonasschnelli** [comment]
@luke-jr "comments" are not directly related with "accounts". By now they where not deprecated. `sendtoaddress` uses them and i think having a possibility to give a transaction a comment makes sense. 

But i'm not really taking a position here. It was listed as a open, unimplemented feature via #2086. I'm more the to-get-a-job-done guy. :)

**Comment by luke-jr** [comment]
Hm, interesting. Surprised that isn't using the label stuff. It would make sense to deprecate labels for comments, in that case, *except* that it's impossible to assign a comment to an incoming request right now. Labels work better in that regard. But having both is just confusing since they serve the same purpose...

**Comment by laanwj** [comment]
Looks like a lot of people are making the label-comment confusion, even in #2086.

Personally I'd say the distinction is the following:

- `label` is a, usually short, name for the recipient or address
- `comment` is arbitrary free text associated the the transaction. This can be anything from a description to e.g. current exchange rates for accounting/tax purposes

The annoying/user-surprising thing about these comments, though, is that they apply to a network-level transaction, so they would seem to be associated with multiple UI transactions if there are multiple outputs.

**Comment by luke-jr** [comment]
It also means you can't add comments to incoming transactions until after they're received (and confirm?). Maybe comment should be a text-area in the UI rather than a single line, and in the new wallet associated with an address so it has the same flexibility?

**Comment by laanwj** [comment]
> Maybe comment should be a text-area in the UI rather than a single line

+1 I somehow assumed that'd already be the case.

**Comment by jonasschnelli** [comment]
> Maybe comment should be a text-area in the UI rather than a single line

This PR will add a multiline text-area for adding comments to outgoing transactions.

>  [...] and in the new wallet associated with an address so it has the same flexibility?

I think adding comments to a wtx is something different to a address label. Lets say if i do ONE wtx to serval recipients it is nice then to give that ONE wtx a comment and not every receiving address.

But i'm also not completely sure what would be best for handling labels and comments.

File: src/qt/walletmodeltransaction.cpp [hunk 2]
```diff
@@ -65,3 +65,8 @@ CReserveKey *WalletModelTransaction::getPossibleKeyChange()
 {
     return keyChange;
 }
+
+void WalletModelTransaction::setComment(const QString comment)
```
Comment (Diapolo) [inline src/qt/walletmodeltransaction.cpp]: You could also use a pass-by-reference here ;)?
``const QString& comment``

**Comment by jonasschnelli** [comment]
rebased and fixed @Diapolo 's nit

**Comment by jgarzik** [comment]
ut ACK

**Comment by jonasschnelli** [comment]
Rebased.

It will slightly reduce the space in the already very packed `sendcoinsdialog`.
But i think it's useful.
I'm not sure if – long term speaking – the Addressbook has a reason to exists. The word "Adressbook" itself is connected to address reuse. If people using the Addressbook to label and track *payments*, then we should slowly remove (or rename and rewrite) the Addressbook and switch over to wallet comments.

**Comment by MarcoFalke** [comment]
Looks good to merge.


screenshot from 2015-09-16 12-10-39

**Comment by gmaxwell** [comment]
utACK. But can we also add an RPC to set these?

**Comment by jonasschnelli** [comment]
@gmaxwell: Thanks for the review. Some test would be required, I agree. I'm also not fully happy with the UI look and feel. Consider the PR as WIP (changed title).

**Comment by pstratem** [comment]
lightly tested ACK 1dcdd428fdee62c1f7fb2e2eff0dc2d63eb0b7f3

**Comment by sipa** [comment]
Concept ACK

**Comment by arowser** [comment]
Can one of the admins verify this patch?

**Comment by paveljanik** [comment]
Concept ACK

RPC first please.

**Comment by jonasschnelli** [comment]
@paveljanik: It's already possible in RPC `sendtoaddress \"bitcoinaddress\" amount ( \"comment\" \"comment-to\" subtractfeefromamount )`

I don't think the current GUI UX makes sense...
Closing for now and maybe reopening after https://github.com/bitcoin/bitcoin/pull/7729

**Comment by paveljanik** [comment]
@jonasschnelli Hmm, it is not possible to use `sendtoaddress` RPC command to add comment to already received transaction.

**Comment by jonasschnelli** [comment]
@paveljanik: this PR only allows to set the comment during GUI tx creation (which is also a point that makes this PR currently not ready).

**Comment by paveljanik** [comment]
Ah, then yes :-)

**Pull request closed without merging**